### PR TITLE
code refactor to optimize the cv application generation

### PR DESCRIPTION
**Pull request by saikumar9**
# PR Checklist

Please check if your PR fulfills the following requirements
- [x] The title follows our [guidelines](https://github.com/ideacrew/enroll/blob/trunk/CONTRIBUTING.md#commit)
- [x] Tests for the changes have been added (for bugfixes/features), they use let helpers and before blocks
- [ ] For all UI changes, there is cucumber coverage
- [ ] Any endpoint touched in the PR has an appropriate Pundit policy. For open endpoints, reasoning is documented in PR and code
- [ ] Any endpoint modified in the PR only responds to the expected MIME types.
- [ ] For all scripts or rake tasks, how to run it is documented on both the PR and in the code
- [ ] There are no inline styles added
- [ ] There are no inline javascript added
- [ ] There is no hard coded text added/updated in helpers/views/Javascript. New/updated translation strings do not include markup/styles, unless there is supporting documentation
- [ ] Code does not use .html_safe
- [ ] All images added/updated have alt text
- [x] Doesn’t bypass rubocop rules in any way

# PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature (requires Feature flag)
- [ ] Data fix or migration (inert code, no impact until run)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Dependency updates (e.g., add a new gem or update to a version)

# What is the ticket # detailing the issue?

Ticket: [Jolly Podgers - 188067360](https://www.pivotaltracker.com/story/show/188067360)

# A brief description of the changes

Current behavior: `N/A`

New behavior: Code refactor for optimizing the CV application transformation.

# Feature Flag

For all new feature development, a feature flag is required to control the exposure of the feature to our end users. A feature flag needs a corresponding environment variable to initialize the state of the flag. Please share the name of the environment variable below that would enable/disable the feature and which client(s) it applies to.

Variable name:

- [ ] DC
- [ ] ME

# Additional Context
Include any additional context that may be relevant to the peer review process.

### CODE REVIEW

File: components/financial_assistance/app/controllers/financial_assistance/application_controller.rb [hunk 1]
```diff
@@ -43,10 +43,12 @@ def support_text_placeholders(raw_support_text)
       end
     end
 
-    # Checks if the application is a renewal draft and if so, sets a flash message and redirects to the applications path.
+    # Checks if the application is a renewal draft and, if so, sets a flash message and redirects to the applications path.
     #
     # @return [void]
     def check_for_uneditable_application
+      # If the application is a renewal draft, the user should not be able to edit it.
+      # The below line could be refactored to read from a constant or a setting in the future if we want to add more states where the application should not be editable.
```
Comment (mdkaraman) [inline components/financial_assistance/app/controllers/financial_assistance/application_controller.rb:51]: Great note here about possible future use cases!

File: components/financial_assistance/app/models/financial_assistance/applicant.rb [hunk 2]
```diff
@@ -178,6 +178,10 @@ class Applicant # rubocop:disable Metrics/ClassLength TODO: Remove this
     field :family_member_id, type: BSON::ObjectId
     field :eligibility_determination_id, type: BSON::ObjectId
 
+    # @deprecated This field is implemented to soft delete the applicant from the system just like the family member.
```
Comment (mdkaraman) [inline components/financial_assistance/app/models/financial_assistance/applicant.rb:181]: Small note-- may be more clear to say "This field `was` implemented to soft delete the applicant from the system just like the family member.  However, the applicant can `now` be deleted from the system as it is only allowed for draft applications, so we are not using this field anymore."

File: components/financial_assistance/app/models/financial_assistance/application.rb [hunk 3]
```diff
@@ -1662,35 +1728,41 @@ def unset_submit
...
     end
 
-    def create_eligibility_determinations
-      ## Remove  when copy method is fixed to exclude copying Tax Household
-      active_applicants.update_all(eligibility_determination_id: nil)
-
-      non_tax_dependents = active_applicants.where(is_claimed_as_tax_dependent: false)
-      tax_dependents = active_applicants.where(is_claimed_as_tax_dependent: true)
-
+    # Builds eligibility determinations for applicants.
```
Comment (mdkaraman) [inline components/financial_assistance/app/models/financial_assistance/application.rb:1731]: Awesome job adding the comments here and removing the unnecessary DB hits.

File: components/financial_assistance/app/models/financial_assistance/application.rb [hunk 4]
```diff
@@ -1640,14 +1693,27 @@ def verification_update_for_applicants
       end
     end
 
+    # Sets the submission details and calculates necessary values for the application.
+    #
+    # @note This method sets data and builds objects, hitting the database only once instead of multiple times for each method.
+    #
+    # @return [void]
     def set_submit
```
Comment (mdkaraman) [inline components/financial_assistance/app/models/financial_assistance/application.rb:1701]: Totally not necessary here, but a thought for the future: we could have a similar clean up and optimization done for the `unset_submit`method.

**Review by mdkaraman — APPROVED** [review]
Great optimizations and enhancements to the CV generation process for FAA Applications.  Significantly reduces the number of times the database is hit when submitting an application and generating the CV payload. Adds a ton of helpful documentation comments as well.  I also noticed this refactor has helped simplify some of the related specs.  

I added a comment/question or two, but nothing that is a blocker, so approving now!

File: components/financial_assistance/app/models/financial_assistance/applicant.rb (hunk 2 above)
Comment (saikumar9) [reply components/financial_assistance/app/models/financial_assistance/applicant.rb:181]: I will update the comment in the next PR.

File: components/financial_assistance/app/models/financial_assistance/application.rb (hunk 4 above)
Comment (saikumar9) [reply components/financial_assistance/app/models/financial_assistance/application.rb:1701]: Yes. We must optimize the `unset_submit ` method. I am planning to create a new ticket for this.

**Review by ymhari — APPROVED** [review]

**Pull request merged**